main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -2377,12 +2436,32 @@ object Resolver {
   /**
     * Returns the type `And(tpe1, tpe2)`.
     */
-  private def mkAnd(tpe1: Type, tpe2: Type, loc: SourceLocation): Type = Type.Apply(Type.Apply(Type.Cst(TypeConstructor.And, loc), tpe1, loc), tpe2, loc)
+  private def mkAnd(tpe1: Type, tpe2: Type, loc: SourceLocation): Type = Type.mkApply(Type.Cst(TypeConstructor.And, loc), List(tpe1, tpe2), loc)
```

Comment:
Any reason these do not call `Type.mkAnd` etc?

Reply:
The helpers in `Type.scala` are too smart. They simplify the type and we lose information that is important for kind inference and error messages.

Reply:
OK, oversmart, if you will.